File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        int n = s.length();
        if(n%2 != 0)return false;
        for(int i = 0; i<n; i++){
            if(s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '['){
                st.push(s.charAt(i));
            }else if(st.size() > 0){
                if(st.peek() == '('){
                    if(s.charAt(i) == ')')
                        st.pop();
                    else
                        return false;
                }else if(st.peek() == '{'){
                    if(s.charAt(i) == '}')
                        st.pop();
                    else
                        return false;
                }else if(st.peek() == '['){
                    if(s.charAt(i) == ']')
                        st.pop();
                    else
                        return false;
                }
            }else{
                return false;
            }
        }
        
        return (st.size() == 0) ? true : false;
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        int n = s.length();
        for(int i=0; i<n; i++){
            if(s[i] == '(' || s[i] == '[' || s[i] == '{'){
                st.push(s[i]);
            }else if(!st.empty() && (s[i] == ')' || s[i] == ']' || s[i]  == '}')){
                if((st.top() == '(' &&  s[i] == ')') || (st.top() == '[' &&  s[i] == ']') || (st.top() == '{' &&  s[i] == '}')){
                    st.pop();
                }else{
                    return false;
                }
            }else{
                return false;
            }
        }
        
        if(st.empty())return true;
        return false;
    }
};